File: grammar_parser/C/parser.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "parser.h"

static FILE* input;
static int la; /* This is a single character lookahead */

/**
  Prints an error message and terminates program execution with an error code.
*/
static void fatal_error(const char* msg)
{
  fprintf(stderr,"Error: %s - Aborting!\n",msg);
  exit(1);
}

/**
  Reads the next character from the input into the lookahead variable.
*/
static void next(void)
{
  la = fgetc(input);
}

/**
  Skips spaces. On return the lookahead is a non-space character or EOF.
*/
void skip_spaces(void)
{
  while (isspace(la)) {
    next();
  }
}

/**
  Expects the lookahead to be one of the characters in matches. It consumes the
  lookahead.
  If lookahead does not match, a fatal error is raised.
*/
static void match(const char* matches)
{
  if (strchr(matches,la)==NULL) {
    fatal_error("Unexpected character");
  }
  next();
}

/**
  Opens the input file filename and initializes the parser.
  Raises a fatal error if the file could not be opened.
*/
static void open_grammar(const char* filename)
{
  input = fopen(filename,"r");
  if (input==NULL) {
    fatal_error("Cannot open input file");
  }
  next();
  skip_spaces();
}

static void close_grammar(void)
{
  fclose(input);
  input = NULL;
}

/*************************************************************************
                    PUT THE GRAMMAR RULES AFTER HERE
**************************************************************************/


/*************************************************************************/

/**
  Parses the input file filename. If the input has sytax errors, an error message is
  printed and the program is terminated with an error code.
*/
void parse_grammar(const char* filename)
{
  open_grammar(filename);
  close_grammar();
}

File: grammar_parser/Java/Parser.java
import java.io.InputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.FileNotFoundException;

class Parser {
  private int la; /* One character lookahead */
  InputStream input;

  /**
    Read next input character into lookahead. la = -1 indicates end of file.
  */
  private void next() {
    try {
      la = input.read();
    }
    catch (IOException e) {
      la = -1;
    }
  }

  private void match(String matches) throws SyntaxErrorException {
    if (matches.indexOf(la)>=0) {
      next();
    }
    else {
      throw new SyntaxErrorException("Unexpected character");
    }
  }

  /**
    Skips whitespace characters until a non-whitespace character or end of file is reached.
  */
  private void skipSpaces() {
    while (Character.isWhitespace(la)) {
      next();
    }
  }

  public Parser(InputStream input) {
    this.input = input;
    next();
    skipSpaces();
  }

  /* ******************************************************************** */
  /*                     PUT YOUR PARSER RULES HERE                       */



  /* ******************************************************************** */

  public void parse_grammar() throws SyntaxErrorException {
    /* Calls your start rule here */
  }

  public static void main(String[] args) {
    if (args.length==1) {
      try {
        FileInputStream input = new FileInputStream(args[0]);
        Parser parser = new Parser(input);
        parser.parse_grammar();
      }
      catch (FileNotFoundException e) {
        System.err.println("Cannot open file "+args[0]);
      }
      catch (SyntaxErrorException e) {
        System.err.println("Syntax Error: "+e.getMessage());
      }
    }
    else {
      System.err.println("Usage: Parser FILENAME");
    }
  }
}
